File: public/python/084-kth-largest-element.py
# LeetCode Problem #215: Kth Largest Element in an Array
# Difficulty: Medium
# Link: https://leetcode.com/problems/kth-largest-element-in-an-array/

from typing import List
import heapq
import random

# ─────────────────────────────────────────────
# APPROACH 1: Min Heap of Size K | O(n log k) time | O(k) space
# EXPLAIN: Maintain a min-heap of exactly k elements; the heap top is always the kth largest.
# WHEN: Best when k is much smaller than n.

def solve_1(nums: List[int], k: int) -> int:
    min_heap = []
    for num in nums:
        heapq.heappush(min_heap, num)
        if len(min_heap) > k:
            heapq.heappop(min_heap)
    return min_heap[0]

# ─────────────────────────────────────────────
# APPROACH 2: Sort Descending | O(n log n) time | O(1) space
# EXPLAIN: Sort in descending order and return the element at index k-1.
# WHEN: When simplicity matters and n is small.

def solve_2(nums: List[int], k: int) -> int:
    nums.sort(reverse=True)
    return nums[k - 1]

# ─────────────────────────────────────────────
# APPROACH 3: Quickselect | O(n) average time | O(log n) space
# EXPLAIN: Partition around a random pivot; recurse only on the side containing the target index.
# WHEN: Best average-case performance; preferred in interviews for O(n) guarantee.

def solve_3(nums: List[int], k: int) -> int:
    target = len(nums) - k  # kth largest = element at index n-k in ascending order

    def partition(lo: int, hi: int) -> int:
        pivot_idx = random.randint(lo, hi)
        nums[pivot_idx], nums[hi] = nums[hi], nums[pivot_idx]
        pivot = nums[hi]
        i = lo
        for j in range(lo, hi):
            if nums[j] <= pivot:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
        nums[i], nums[hi] = nums[hi], nums[i]
        return i

    def quickselect(lo: int, hi: int) -> int:
        if lo == hi:
            return nums[lo]
        p = partition(lo, hi)
        if p == target:
            return nums[p]
        elif p < target:
            return quickselect(p + 1, hi)
        else:
            return quickselect(lo, p - 1)

    return quickselect(0, len(nums) - 1)

# ─────────────────────────────────────────────
# APPROACH 4: Max Heap (Extract k Times) | O(n + k log n) time | O(n) space
# EXPLAIN: Build a max-heap from all elements, then pop k times to reach the kth largest.
# WHEN: When k is close to n, or when you want a simple heap-based solution.

def solve_4(nums: List[int], k: int) -> int:
    max_heap = [-x for x in nums]
    heapq.heapify(max_heap)
    for _ in range(k - 1):
        heapq.heappop(max_heap)
    return -max_heap[0]

# ─────────────────────────────────────────────
# APPROACH 5: Counting Sort | O(n + R) time | O(R) space
# EXPLAIN: Use a frequency array over the bounded value range [-10^4, 10^4] and scan from high to low.
# WHEN: When values are bounded in a known small range (as guaranteed by constraints).

def solve_5(nums: List[int], k: int) -> int:
    OFFSET = 10000
    R = 20001
    cnt = [0] * R
    for n in nums:
        cnt[n + OFFSET] += 1
    remain = k
    for v in range(R - 1, -1, -1):
        remain -= cnt[v]
        if remain <= 0:
            return v - OFFSET
    return -1  # unreachable

# Made with Bob

File: public/java/084-kth-largest-element.java
/*
 * LeetCode Problem #215: Kth Largest Element in an Array
 * Link: https://leetcode.com/problems/kth-largest-element-in-an-array/
 * Difficulty: Medium
 */
import java.util.*;

class Solution {

    // APPROACH 1: Min Heap of Size K | O(n log k) time | O(k) space
    // EXPLAIN: Maintain a min-heap of exactly k elements; the top is the kth largest.
    public int findKthLargest1(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) minHeap.poll();
        }
        return minHeap.peek();
    }

    // APPROACH 2: Sort Descending | O(n log n) time | O(1) space
    // EXPLAIN: Sort in descending order and return the element at index k-1.
    public int findKthLargest2(int[] nums, int k) {
        Arrays.sort(nums);
        return nums[nums.length - k];
    }

    // APPROACH 3: Quickselect | O(n) average time | O(log n) space
    // EXPLAIN: Partition array in descending order around a pivot; recurse only on the relevant side.
    public int findKthLargest3(int[] nums, int k) {
        return quickselect(nums, 0, nums.length - 1, k - 1);
    }

    private int quickselect(int[] nums, int lo, int hi, int target) {
        if (lo == hi) return nums[lo];
        int p = partition(nums, lo, hi);
        if (p == target) return nums[p];
        else if (p < target) return quickselect(nums, p + 1, hi, target);
        else return quickselect(nums, lo, p - 1, target);
    }

    private int partition(int[] nums, int lo, int hi) {
        int pivot = nums[hi], i = lo;
        for (int j = lo; j < hi; j++) {
            if (nums[j] >= pivot) { // descending order
                int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp;
                i++;
            }
        }
        int tmp = nums[i]; nums[i] = nums[hi]; nums[hi] = tmp;
        return i;
    }

    // APPROACH 4: Max Heap (Extract k Times) | O(n + k log n) time | O(n) space
    // EXPLAIN: Build a max-heap from all elements, then pop k times.
    public int findKthLargest4(int[] nums, int k) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for (int num : nums) maxHeap.offer(num);
        for (int i = 0; i < k - 1; i++) maxHeap.poll();
        return maxHeap.peek();
    }

    // APPROACH 5: Counting Sort | O(n + R) time | O(R) space
    // EXPLAIN: Use a frequency array over the bounded range [-10^4, 10^4] and scan from high to low.
    public int findKthLargest5(int[] nums, int k) {
        int OFFSET = 10000, R = 20001;
        int[] cnt = new int[R];
        for (int n : nums) cnt[n + OFFSET]++;
        int remain = k;
        for (int v = R - 1; v >= 0; v--) {
            remain -= cnt[v];
            if (remain <= 0) return v - OFFSET;
        }
        return -1;
    }
}

// Made with Bob
